ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/Snapshots.hs

```diff
@@ -180,32 +194,99 @@ diskSnapshotIsPermanent = isJust . dsSuffix
 diskSnapshotIsTemporary :: DiskSnapshot -> Bool
 diskSnapshotIsTemporary = not . diskSnapshotIsPermanent
 
--- | Read snapshot from disk
+data ReadSnaphotErr =
+    -- | Error while de-serialising data
+    ReadSnaphotFailed ReadIncrementalErr
+    -- | Checksum of read snapshot differs from the one tracked by @'DiskSnapshot'@
+  | ReadSnaphotDataCorruption
+    -- | A CRC file does not exist for a @'DiskSnapshot'@
+  | ReadSnapshotNoCRC
+    -- | A CRC file exists for a @'DiskSnapshot'@, but it's contents is invalid
+  | ReadSnapshotBadCRC
+  deriving (Eq, Show)
+
+-- | Read snapshot from disk.
+--
+--   Fail on data corruption, i.e. when the checksum of the read data differs
+--   from the one tracked by @'DiskSnapshot'@.
 readSnapshot ::
      forall m blk. IOLike m
   => SomeHasFS m
   -> (forall s. Decoder s (ExtLedgerState blk))
   -> (forall s. Decoder s (HeaderHash blk))
   -> DiskSnapshot
-  -> ExceptT ReadIncrementalErr m (ExtLedgerState blk)
-readSnapshot hasFS decLedger decHash =
-      ExceptT
-    . readIncremental hasFS decoder
-    . snapshotToPath
+  -> ExceptT ReadSnaphotErr m (ExtLedgerState blk)
+readSnapshot someHasFS decLedger decHash snapshotName = do
+  snapshotCRC <- readCRC someHasFS (snapshotToChecksumPath snapshotName)
+  ExceptT $
+    readIncremental someHasFS decoder (snapshotToPath snapshotName) >>= \case
+       Left e  -> pure $ Left (ReadSnaphotFailed e)
+       Right (ledgerState, checksumAsRead) ->
+           if checksumAsRead /= snapshotCRC
+           then pure $ Left ReadSnaphotDataCorruption
+           else pure (Right ledgerState)
```

Comment:
Perhaps this could be written inside `ExceptT`?

```suggestion
  (ledgerState, checsumAsRead) <- 
    withExceptT 
      ReadSnapshotFailed 
      (readIncremental someHasFS decoder (snapshotToPath snapshotName))
  if checksumAsRead /= snapshotCRC
   then throwE ReadSnaphotDataCorruption
   else pure ledgerState
```